### PR TITLE
Add dynamic port allocation

### PR DESCRIPTION
**Pull request by cwdsuzhou**

### CODE REVIEW

File: pkg/controllers/gameservers/controller.go [hunk 1]
```diff
@@ -369,13 +393,69 @@ func (c *Controller) syncGameServerDeletionTimestamp(gs *carrierv1alpha1.GameSer
...
+	}
+	allocateType := getAllocateType(gs)
+	// only allow use multi ports or directly use port range.
+	gsCopy := gs.DeepCopy()
+	var (
+		ports []int
+		err   error
+	)
+	switch allocateType {
+	case "":
```
Comment (hustcat) [inline pkg/controllers/gameservers/controller.go:415]: For `LoadBalancer` policy, or `Spec.Ports == nil `, we can skip this.

**Review by hustcat — CHANGES_REQUESTED** [review]

File: pkg/controllers/gameservers/controller.go (hunk 1 above)
Comment (cwdsuzhou) [reply pkg/controllers/gameservers/controller.go:415]: Done

File: pkg/controllers/gameservers/controller.go [hunk 2]
```diff
@@ -369,13 +393,72 @@ func (c *Controller) syncGameServerDeletionTimestamp(gs *carrierv1alpha1.GameSer
 	return gs, errors.Wrap(err, "error removing finalizer for GameServer")
 }
 
+// tryAllocatePorts try to allocate port for GameServer, whose port policy is Dynamic.
+func (c *Controller) tryAllocatePorts(gs *carrierv1alpha1.GameServer) (*carrierv1alpha1.GameServer, error) {
+	if len(gs.Spec.Ports) == 0 {
+		return gs, nil
```
Comment (cwdsuzhou) [inline pkg/controllers/gameservers/controller.go:399]: Spec.Ports == nil  checked here

**Review by hustcat — APPROVED** [review]
/lgtm

**Pull request merged**